pkg/agent/flowexporter/flowrecords/flow_records.go

```diff
@@ -15,92 +15,86 @@
 package flowrecords
 
 import (
+	"fmt"
 	"net"
+	"sync"
+	"time"
 
 	"k8s.io/klog"
 
 	"github.com/vmware-tanzu/antrea/pkg/agent/flowexporter"
-	"github.com/vmware-tanzu/antrea/pkg/agent/flowexporter/connections"
 )
 
 type FlowRecords struct {
 	// Add lock when this map is consumed by more than one entity concurrently.
 	recordsMap map[flowexporter.ConnectionKey]flowexporter.FlowRecord
-	connStore  *connections.ConnectionStore
+	mutex      sync.Mutex
 }
 
-func NewFlowRecords(connStore *connections.ConnectionStore) *FlowRecords {
+func NewFlowRecords() *FlowRecords {
 	return &FlowRecords{
-		make(map[flowexporter.ConnectionKey]flowexporter.FlowRecord),
-		connStore,
+		recordsMap: make(map[flowexporter.ConnectionKey]flowexporter.FlowRecord),
 	}
 }
 
 // BuildFlowRecords builds the flow record map from connection map in connection store
-func (fr *FlowRecords) BuildFlowRecords() error {
-	addOrUpdateFlowRecord := func(key flowexporter.ConnectionKey, conn flowexporter.Connection) error {
-		// If DoExport flag is not set return immediately.
-		if !conn.DoExport {
-			return nil
-		}
-		record, exists := fr.recordsMap[key]
+func (fr *FlowRecords) AddOrUpdateFlowRecord(key flowexporter.ConnectionKey, conn flowexporter.Connection) error {
+	fr.mutex.Lock()
+	defer fr.mutex.Unlock()
 
+	record, exists := fr.recordsMap[key]
+	if !exists {
 		isIPv6 := false
 		if net.ParseIP(key[0]).To4() == nil {
 			isIPv6 = true
 		}
-		if !exists {
-			record = flowexporter.FlowRecord{
-				Conn:               &conn,
-				PrevPackets:        0,
-				PrevBytes:          0,
-				PrevReversePackets: 0,
-				PrevReverseBytes:   0,
-				IsIPv6:             isIPv6,
-			}
-		} else {
-			record.Conn = &conn
+		record = flowexporter.FlowRecord{
+			Conn:               &conn,
+			PrevPackets:        0,
+			PrevBytes:          0,
+			PrevReversePackets: 0,
+			PrevReverseBytes:   0,
+			IsIPv6:             isIPv6,
+			LastExportTime:     conn.StartTime,
+			IsActive:           true,
```

Comment:
If there was an idle timeout for a flow, the flow record object was deleted, which as I pointed out previously I agree with. However, if the connection becomes active again, it seems that:
* we use the connection counters (which have not been reset) for current counts
* The "prev" counters are initialized to 0
So won't we report big delta in the first flow record when the connection becomes active again?

Reply:
We need to agree on the solution:
* consider that it is part of the same IPFIX flow but that means we need to preserve "prev" counts, or
* consider that it is a new IPFIX flow (obviously it's part of the same TCP flow), and use a new startTime and reset counters

Reply:
and we would need a unit test for that case as well, I don't think there is one currently

Reply:
>>we use the connection counters (which have not been reset) for current counts
>>The "prev" counters are initialized to 0
>>So won't we report big delta in the first flow record when the connection becomes active again?

When "previous" counters are 0, we send [delta counters as 0](https://github.com/vmware-tanzu/antrea/blob/main/pkg/agent/flowexporter/exporter/exporter.go#L376). This is true for the first flow record of new IPFIX flow. 
Therefore, when the new IPFIX flow is created for the existing connection, delta counters are 0 for the first flow record and the previous stats are updated once the first record is sent. Deltas are computed for subsequent flow records and we will not be sending large deltas.
startTime for the new IPFIX flow is the startTime of the existing connection and not a new startTime. I think this is ok. Do you see any issue here?

Will add a unit test for this.

Reply:
Then the only issue I see is that we don't start a new IPFIX flow, yet it seems that over the lifetime of the flow, the sum of deltas is not equal to the total count, unless I misunderstand your comment:

1) TCP flow created, 100 packets are sent
2) flow becomes idle - send report with deltaCnt = 100, totalCnt = 100
3) flow resumes, 100 packets are sent
4) send report with deltaCnt = 0, totalCnt = 200

I would say that my understanding is that if you purge your flowMap, you should consider it a new IPFIX flow and reset the startTime. If this is not the right thing for us, maybe we should not purge the flowMap?

Reply:
Addressed this comment by not purging the flow record from the cache as soon as it becomes idle. We keep the flow record as long as the connection is there in the conntrack table, and resulting flow records will have consistent stats. For idle connections, we just send one flow record with idleTimeout and check if the connection becomes active again through stats (packet counts) to send flow record with activeTimeout. 

Waiting on a [fix in go-ipfix](https://github.com/vmware/go-ipfix/pull/139) to add unit tests for flowExporter.Export method to test the above scenario and other regular export cases.